Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,271 @@
+        {
+            BattleId = battleId,
+            Fee = fee,
+            Users = new Address[MaxUsers]
+        };
+        SetBattle(battleId, battle);
+
+        Log(new BattleEventLog { Event = "Start", BattleId = battleId, Address = Message.Sender });
+        return battleId;
+    }
+    /// <summary>
+    /// 4 different user will enter the battle
+    /// </summary>
+    public void EnterBattle(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle not found.");
+
+        Assert(battle.Fee == Message.Value, "Battle fee is not matching with entry fee paid.");
+
+        var user = GetUser(battleId, Message.Sender);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        SetUser(battleId, Message.Sender, user);
+
+        var userindex = GetUserIndex(battleId);
+        Assert(userindex != MaxUsers, "Max user reached for this battle.");
+        battle.Users.SetValue(Message.Sender, userindex);
+        SetUserIndex(battleId, (userindex + 1));
+
+        SetBattle(battleId, battle);
+
+        Log(new BattleEventLog { Event = "Enter", BattleId = battleId, Address = Message.Sender });
+    }
+    /// <summary>
+    /// 4 different user will end the battle and submit the score
+    /// </summary>
+    public void EndBattle(Address userAddress, ulong battleId, uint score)
+    {
+        Assert(Message.Sender == Owner, "Only battle owner can end game.");
+
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle not found.");
+
+        var user = GetUser(battleId, userAddress);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        var ScoreSubmittedCount = GetScoreSubmittedCount(battleId);
+        ScoreSubmittedCount += 1;
+        SetScoreSubmittedCount(battleId, ScoreSubmittedCount);
+
+        var highestScorer = GetHighestScorer(battleId);
+
+        if (score > highestScorer.Score)
+            SetHighestScorer(battleId, new BattleHighestScorer { Scorers = new Address[] { userAddress }, Score = score });
+        else if (score == highestScorer.Score)
+        {
+            var scorers = highestScorer.Scorers;
+            Array.Resize(ref scorers, scorers.Length + 1);
+            scorers[scorers.Length - 1] = userAddress;
+            SetHighestScorer(battleId, new BattleHighestScorer { Scorers = scorers, Score = score });
+        }
+
+        if (ScoreSubmittedCount == MaxUsers)
+        {
+            highestScorer = GetHighestScorer(battleId);
+            if (highestScorer.Scorers.Length > 1)
+                CancelBattle(battle);
+            else
+                ProcessWinner(battle, highestScorer.Scorers[0]);
+        }
+
+        Log(new BattleEventLog { Event = "End", BattleId = battleId, Address = Message.Sender });
```

Comment:
It is usually recommended way to log by type name like `BattleEndedLog` instead of flag property.